wallet-integration/guide.mdx

```diff
@@ -30,24 +30,108 @@ To achieve this goal, you will need to do the following:
 
 Let's dive deeper into each of these steps:
 
-## Detecting a new website
+## Detecting navigation to a new website
+
+<Note>
+  In the code examples below, we're using Mozilla's
+  [webextension-polyfill](https://github.com/mozilla/webextension-polyfill) to
+  smooth out the differences between the Chrome and Firefox APIs.
+</Note>
 
 The first step is to detect when a user has visited a new website. If you're building a
-web browser extension, you can use the `browser.onBeforeNavigate` API. Here
-we're using Mozilla's [webextension-polyfill](https://github.com/mozilla/webextension-polyfill)
-to smooth out the differences between the Chrome and Firefox APIs.
+web browser extension, you have a couple of choices:
+
+### Option 1: Use the `browser.onBeforeNavigate` API
+
+This approach requires adding the `webNavigation` permission to your `manifest.json` file:
+
+```json
+{
+  "permissions": ["webNavigation"]
+}
+```
+
+<Warning>
+
+If your extension does not already have this permission, this will trigger a new
+permissions request when the user installs your extension to "Read your browsing history".
+
+Alternatively, you can make this permission optional by adding the
+`"optional_permissions": ["webNavigation"]` key to your `manifest.json` file instead.
+But this will require you to add additional logic to your extension to request this
+permission at the appropriate time (ex. in a settings page, or via a Call to Action
+prompt to opt-in to protection). You will also need to handle the case where the user
+denies the permission request.
+
+</Warning>
+
+Then you can listen for the `onBeforeNavigate` event in your Background script:
 
 ```ts
 import browser from "webextension-polyfill";
 
 browser.webNavigation.onBeforeNavigate.addListener(async function (details) {
-  //do stuff
+  // Filter out subframes and any navigation events that don't have a URL
+  if (details.frameId !== 0 || !details.url) {
+    return;
+  }
+
+  // ...
+});
+```
+
+### Option 2: Use the `browser.tabs.onUpdated` API
+
+This approach requires adding the `tabs` permission to your `manifest.json` file
+and requesting the `"<all_urls>"` host permission:
+
+```json
+{
+  "permissions": ["tabs"],
+  "host_permissions": ["<all_urls>"]
+}
+```
+
+<Warning>
+
+If your extension does not already have this permission, this will trigger a new
```

Comment:
Does it trigger this warning? I thought tabs / active tab might be able to get around showing a warning?

Reply:
I'm still a bit iffy about this aspect. What I found is that on fresh installs, this is what you get:

If I add just `webNavigation` it says it wants to read your browsing history:

| With `webNavigation`                                                                                                                                       | Without `webNavigation`                                                                                                                                    |
| --------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <img width="478" alt="CleanShot 2023-08-22 at 20 06 25@2x" src="https://github.com/chainpatrol/docs/assets/8302959/12891822-317e-42c6-8476-b99bcd31ab4b"> | <img width="515" alt="CleanShot 2023-08-22 at 20 03 32@2x" src="https://github.com/chainpatrol/docs/assets/8302959/e216318d-3515-4fa6-b413-5e2fef2fd0ce"> |


If I add just `tabs` it says it wants to read your browsing history:

| With `tabs`                                                                                                                                       | Without `tabs`                                                                                                                                    |
| --------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <img width="470" alt="CleanShot 2023-08-22 at 20 09 51@2x" src="https://github.com/chainpatrol/docs/assets/8302959/08069942-2162-4ee4-8647-5f3066c8688c"> | <img width="515" alt="CleanShot 2023-08-22 at 20 03 32@2x" src="https://github.com/chainpatrol/docs/assets/8302959/e216318d-3515-4fa6-b413-5e2fef2fd0ce"> |


If I add `tabs` + `"host_permissions": ["<all_urls>"]` it says it wants to read and change data on all websites:

| With `tabs` + `host_permissions`                                                                                                                                      | Without `tabs` + `host_permissions`                                                                                                                                    |
| --------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <img width="472" alt="CleanShot 2023-08-22 at 20 11 29@2x" src="https://github.com/chainpatrol/docs/assets/8302959/788771c6-d04d-4dea-ac0b-1e2599e95e14"> | <img width="515" alt="CleanShot 2023-08-22 at 20 03 32@2x" src="https://github.com/chainpatrol/docs/assets/8302959/e216318d-3515-4fa6-b413-5e2fef2fd0ce"> |

With no permissions, but if you have a content script that has `<all_urls>` it will trigger this warning:

| With content script                                                                                                                                       | Without content script                                                                                                                                    |
| --------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <img width="488" alt="CleanShot 2023-08-22 at 19 59 42@2x" src="https://github.com/chainpatrol/docs/assets/8302959/1832d52b-29ea-4582-83c5-3cea00feb4a5"> | <img width="515" alt="CleanShot 2023-08-22 at 20 03 32@2x" src="https://github.com/chainpatrol/docs/assets/8302959/e216318d-3515-4fa6-b413-5e2fef2fd0ce"> |

So I feel like the first two options are the best. 

Also, I think I misinterpreted the docs [here](https://developer.chrome.com/docs/extensions/mv3/permission_warnings/#overview):

<img width="847" alt="CleanShot 2023-08-22 at 20 20 35@2x" src="https://github.com/chainpatrol/docs/assets/8302959/05464148-1846-4226-82c0-ea266df138e5">

I thought it was saying that it won't show a warning if you have both permissions. But what it's really saying is that `tabs` won't warn if you already have `<all_urls>` because `<all_urls>` is more powerful than just `tabs`.

Regarding `activeTab`, that doesn't really do anything for us unless the interaction is "initiated by the user". This has a [very narrow meaning](https://developer.chrome.com/docs/extensions/mv3/manifest/activeTab/#invoking-activeTab):

<img width="657" alt="CleanShot 2023-08-22 at 20 25 23@2x" src="https://github.com/chainpatrol/docs/assets/8302959/06f45da3-13cc-4d27-9796-9e1b1db2c814">

So if the browser extension is using a popup, this is not a feasible pathway.

I'll update this doc to remove that `host_permissions` part and add a note about using the content scripts too that it will require quite bad permission there